Core/Modes/Matchmode.cs

```diff
@@ -1,43 +1,132 @@
         private readonly ModeBase _modeBase;
+        private readonly WebsocketBroadcastServer _broadcastServer;
+        private readonly OverlayConnection _overlayConnection;
 
         public Matchmode(ILoggerFactory loggerFactory, BaseConfig baseConfig, MatchmodeConfig matchmodeConfig)
         {
             _matchmodeConfig = matchmodeConfig;
+            _loggerFactory = loggerFactory;
             _logger = loggerFactory.CreateLogger<Matchmode>();
             _stopToken = new StopToken();
             _modeBase = new ModeBase(loggerFactory, baseConfig, _stopToken);
+
+            _broadcastServer = new WebsocketBroadcastServer(
+                loggerFactory.CreateLogger<WebsocketBroadcastServer>(), "localhost", 5001);
+            _overlayConnection =
+                new OverlayConnection(loggerFactory.CreateLogger<OverlayConnection>(), _broadcastServer);
         }
 
         public async Task Run()
         {
             _logger.LogInformation("Matchmode starting");
             _modeBase.Start();
+            Task overlayWebsocketTask = _broadcastServer.Listen();
             while (!_stopToken.ShouldStop)
             {
-                // TODO match main loop goes here
-                await Task.Delay(TimeSpan.FromMilliseconds(100));
+                _loopCancelToken = new CancellationTokenSource();
+                try
+                {
+                    await Loop(_loopCancelToken.Token);
+                }
+                catch (OperationCanceledException)
+                {
+                    _logger.LogInformation("Match loop cancelled");
+                }
             }
+            await _broadcastServer.Stop();
+            await overlayWebsocketTask;
             _logger.LogInformation("Matchmode ended");
         }
 
+        private async Task Loop(CancellationToken cancellationToken)
+        {
+            var teams = new Teams
+            {
+                Blue = ImmutableList.Create(MatchTesting.TestVenonatForOverlay),
+                Red = ImmutableList.Create(MatchTesting.TestVenonatForOverlay),
+            };
+            await Task.Delay(TimeSpan.FromSeconds(3), cancellationToken);
+
+            IMatchCycle match = new CoinflipMatchCycle(_loggerFactory.CreateLogger<CoinflipMatchCycle>());
+            Task setupTask = match.SetUp(new MatchInfo(teams.Blue, teams.Red), cancellationToken);
+            await _overlayConnection.Send(new MatchCreatedEvent(), cancellationToken);
+            await _overlayConnection.Send(new MatchBettingEvent(), cancellationToken);
+            await _overlayConnection.Send(new MatchModesChosenEvent(), cancellationToken); // TODO
+            await _overlayConnection.Send(new MatchSettingUpEvent
+            {
+                MatchId = 1234,
+                Teams = teams,
+                BettingDuration = _matchmodeConfig.DefaultBettingDuration.TotalSeconds,
+                RevealDuration = 0,
+                Gimmick = "speed",
```

Comment:
idk if gimicks, switching and battle style as strings is a good idea, I'd rather them be enums

Reply:
Is there even a well-defined list of gimmicks? I think it's just a huge configuration file I'm not familiar with